colossus/src/main/scala/colossus/controller/OutputController.scala

```diff
@@ -408,39 +382,68 @@ trait OutputController[Input, Output] extends MasterController[Input, Output] {
     case state: Alive[Output] => {
       if (checkOutputGracefulDisconnect(state)) {
         MoreDataResult.Complete
-      } else {
-        val fullBefore = state.dataQ.isFull
-        var continue = state.dataQ.itemSize > 0
-        while (continue) {
-          val next = state.dataQ.head
-          buffer.copy(next)
-          if (next.remaining == 0) {
-            state.dataQ.dequeue
-          } else {
-            continue = false
+      } else state.liveState match {
+        case Dequeueing => {
+          var streamState: Option[Streaming] = None
+          var continue = true
+          while (writesEnabled && continue && state.msgQ.size > 0 && ! buffer.isOverflowed) {
+            val next = state.msgQ.dequeue
+            codec.encode(next.item) match {
+              case e: Encoder => {
+                e.encode(buffer)
+                next.postWrite(OutputResult.Success)
+              }
+              case DataStream(source) => {
+                val ls = Streaming(source, new DataQueue(1000), next.postWrite)
+                streamState = Some(ls)
+                outputState = state.copy(liveState = ls)
+                drainSource(ls)
+                continue = false
+              }
+            }
           }
-          if (state.dataQ.itemSize == 0) {
-            continue = false
+          //TODO this can be cleaned up somehow
+          if (continue) {
+            if (state.disconnecting || state.msgQ.size > 0) {
```

Comment:
I've created #323 to address this.  Changing the way `OutputState` works will get rid of a lot of this.